diffrax/brownian/tree.py

```diff
@@ -88,70 +152,224 @@ def __init__(
+        r = eqxi.error_if(
+            r,
+            r < 0,
+            "Cannot evaluate VirtualBrownianTree outside of its range [t0, t1].",
+        )
+        r = eqxi.error_if(
+            r,
+            r > 1,
+            "Cannot evaluate VirtualBrownianTree outside of its range [t0, t1].",
+        )
+        # Clip because otherwise the while loop below won't terminate, and the above
+        # errors are only raised after everything has finished executing.
+        r = jnp.clip(r, 0, 1)
+        map_func = lambda key, shape: self._evaluate_leaf(key, r, shape)
         return jtu.tree_map(map_func, self.key, self.shape)
 
-    def _brownian_bridge(self, s, t, u, w_s, w_u, key, shape, dtype):
-        mean = w_s + (w_u - w_s) * ((t - s) / (u - s))
-        var = (u - t) * (t - s) / (u - s)
-        std = jnp.sqrt(var)
-        return mean + std * jrandom.normal(key, shape, dtype)
+    def _brownian_arch(
+        self,
+        level: int,
+        s: Scalar,
+        u: Scalar,
+        w: Tuple[Scalar, Scalar, Scalar],
+        key,
+        shape,
+        dtype,
+        bhh: Optional[Tuple[Scalar, Scalar, Scalar]],
+        bkk: Optional[Tuple[Scalar, Scalar, Scalar]],
+    ):
+        r"""For `t = (s+u)/2` evaluates `w_t` and (optionally) `bhh_t`
+         conditioned on `w_s`, `w_u`, `bhh_s`, `bhh_u`, where
+         `bhh_st` represents $\bar{H}_{s,t} \coloneqq (t-s) H_{s,t}$.
+         To avoid cancellation errors, requires an input of `w_su`, `bhh_su`
+         and also returns `w_st` and `w_tu` in addition to just `w_t`. Same for `bhh`
+         if it is not None.
+         Note that the inputs and outputs already contain `bkk`. These values are
+         there for the sake of a future extension with "space-time-time" Levy area
+         and should be None for now.
+
+        **Arguments:**
+            - `s`: start time
+            - `u`: end time
+            - `w_s`: value of BM at s
+            - `w_u`: value of BM at u
+            - `w_su`: $W_{s,u}$
+            - `key`:
+            - `shape`:
+            - `dtype`:
+            - `bhh`: (optional) $(\bar{H}_s, \bar{H}_u, \bar{H}_{s,u})$
+            - `bkk`: (optional) $(\bar{K}_s, \bar{K}_u, \bar{K}_{s,u})$
+
+        **Returns:**
+            - `t`: midpoint time
+            - `w_t`: value of BM at t
+            - `w_st_tu`: $(W_{s,t}, W_{t,u})$
+            - `bhh`: (optional) $(\bar{H}_s, \bar{H}_t, \bar{H}_u)$
+            - `bhh_st_tu`: (optional) $(\bar{H}_{s,t}, \bar{H}_{t,u})$
+            - `bkk_t`: (optional) $(\bar{K}_s, \bar{K}_t, \bar{K}_u)$
+            - `bkk_st_tu`: (optional) $(\bar{K}_{s,t}, \bar{K}_{t,u})$
+
+        """
+
+        su = jnp.power(jnp.asarray(2.0, dtype), -level)
+        st = su / 2
+        t = s + st
+        u_minus_s = u - s
+        su = eqxi.error_if(
+            su,
+            jnp.abs(u_minus_s - su) > 1e-17,
+            "VirtualBrownianTree: u-s is not 2^(-tree_level)",
+        )
```

Comment:
I don't think we actually need this test at runtime? IIRC this was just for checking things whilst developing.